src/mqtt/protocol/mqtt/mqtt_client.c

```diff
@@ -0,0 +1,1110 @@
+		// indicating a successful receipt of a QoS 2 message
+		work_set_final(work);
+		nni_id_remove(&p->recv_unack, work->packet_id);
+		// ownership of work->msg to the lmq
+		mqtt_pipe_recv_msgq_putq(p, work->msg);
+		work->msg = NULL;
+		mqtt_sock_free_work(s, work);
+		mqtt_run_recv_queue(s);
+		mqtt_send_start(s);
+		nni_mtx_unlock(&s->mtx);
+		return;
+
+	case NNG_MQTT_PINGREQ:
+		// we have sent a PINGREQ
+		mqtt_send_start(s);
+		nni_mtx_unlock(&s->mtx);
+		return;
+
+	case NNG_MQTT_PUBLISH:
+		// we have sent a PUBLISH
+		if (0 == work->qos) {
+			// QoS 0, no further actions
+			work_set_final(work);
+		} else {
+			if (1 == work->qos) {
+				// for QoS 1, expect receiving a PUBACK
+				if (!work_is_acked(work)) {
+					work_set_recv(work, NNG_MQTT_PUBACK);
+					scheduler = true;
+				} else {
+					// scheduling disorder, PUBACK received
+					work_set_final(work);
+				}
+			} else {
+				// for QoS 2, expect receiving a PUBREC
+				if (!work_is_acked(work)) {
+					work_set_recv(work, NNG_MQTT_PUBREC);
+					scheduler = true;
+				} else {
+					// scheduling disorder, PUBREC received
+					work_set_recv(work, NNG_MQTT_PUBCOMP);
+				}
+			}
+		}
+		break;
+
+	case NNG_MQTT_PUBREC:
+		// we have sent a PUBREC, expect receiving a PUBREL
+		work_set_recv(work, NNG_MQTT_PUBREL);
+		break;
+
+	case NNG_MQTT_PUBREL:
+		// we have sent a PUBREL, expect receiving a PUBCOMP
+		work_set_recv(work, NNG_MQTT_PUBCOMP);
+		break;
+
+	case NNG_MQTT_SUBSCRIBE:
+		// we have sent a SUBSCRIBE, expect receiving a SUBACK
+		if (!work_is_acked(work)) {
+			work_set_recv(work, NNG_MQTT_SUBACK);
+			scheduler = true;
+		} else {
+			// scheduling disorder, SUBACK received
+			work_set_final(work);
+		}
+		break;
+
+	case NNG_MQTT_UNSUBSCRIBE:
+		// we have sent a UNSUBSCRIBE, expect receiving a UNSUBACK
+		if (!work_is_acked(work)) {
+			work_set_recv(work, NNG_MQTT_UNSUBACK);
+			scheduler = true;
+		} else {
+			// scheduling disorder, UNSUBACK received
+			work_set_final(work);
+		}
+		break;
+
+	default:
+		work_set_error(work);
+		break;
+	}
+
+	if (work_is_error(work)) {
+		// MQTT protocol error, terminate the connection
+		nni_aio_finish_error(work->user_aio, NNG_EPROTO);
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	} else if (work_is_final(work)) {
+		// good news, protocol state machine run to the end
+		nni_aio *aio = work->user_aio;
+		mqtt_sock_free_work(s, work);
+		mqtt_send_start(s);
+		nni_mtx_unlock(&s->mtx);
+		if (NULL != aio) {
+			nni_aio_finish(aio, 0, 0);
+		}
+		return;
+	}
+
+	mqtt_send_start(s);
+	nni_mtx_unlock(&s->mtx);
+	if (scheduler) {
+		work_timer_schedule(work);
+	}
+	return;
+}
+
+static void
+mqtt_recv_cb(void *arg)
+{
+	mqtt_pipe_t *p = arg;
+	mqtt_sock_t *s = p->mqtt_sock;
+	work_t *     work;
+
+	nni_mtx_lock(&s->mtx);
+
+	nni_msg *msg = nni_aio_get_msg(&p->recv_aio);
+
+	if (nni_aio_result(&p->recv_aio) != 0) {
+		nni_msg_free(msg);
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
```

Comment:
nni_pipe_ etc. are probably going to be reassociated with SP.  The details of transports are fairly SP specific, and I really want to move this work off of using the SP methods.  I know that we have some work to do to finish enabling that.  Just be advised that this will need to change in the future.

Reply:
roger that. I am using MQTT specific transport layer for now.